### ISSUE TITLE
add tests for AutoNAT

### ISSUE THREAD

**Issue by marten-seemann** [libp2p/test-plans#124]
We should have a test that checks that nodes can reliably determine if they are behind a NAT. This will most likely require us to implement some kind of NAT (equivalent?), maybe using  `iptables`.

Main challenge: 
* Implementing the NAT. 
* AutoNAT requires a few interactions with different nodes, so we’d have to boot up a small network of helper nodes.

**Comment by MarcoPolo** [comment]
Should we just focus on autonat v2?

**Comment by marten-seemann** [comment]
Makes sense. Having this tested end-to-end (with "real" NATs) is valuable, even if rust-libp2p and js-libp2p don't have an implementation right away.

**Comment by p-shahi** [comment]
No significant users of AutoNat in rust-libp2p, js-libp2p has this feature but unsure of users.
Although only one implementation will have v2 (go-libp2p), it will still be valuable to add as an interop test as it's difficult to test in unit tests

**Comment by thomaseizinger** [comment]
What is the progress on the V2 spec and the go implementation? I am interested in having V2 asap in `rust-libp2p` because with our recent release of kademlia client-mode, knowing our external addresses has become more important.

**Comment by sukunrt** [comment]
I'm busy finishing the smart dialing effort. https://github.com/libp2p/go-libp2p/issues/2394 is the last major piece left. I will pick AutoNAT v2 implementation after next week. I'll start addressing the open issues on the specs from Monday. 

So far as I understand there are two major open issues on the specs. 
1. How to ensure that we don't reuse the socket that was used to make the dial back request. https://github.com/libp2p/specs/pull/538#discussion_r1201206137 
2. How should we think about amplification attack prevention? https://github.com/libp2p/specs/pull/538#discussion_r1245433030

**Comment by thomaseizinger** [comment]
> I'm busy finishing the smart dialing effort. [libp2p/go-libp2p#2394](https://github.com/libp2p/go-libp2p/issues/2394) is the last major piece left. I will pick AutoNAT v2 implementation after next week. I'll start addressing the open issues on the specs from Monday.

Cool! I didn't mean to stress by the way, just asking! :)
Looking forward to the spec discussions.
